File: 9663.py
import sys


def search(k):
    global cnt
    if k == N:
        cnt += 1
    else:
        for i in range(N):
            if col[i] | diag1[i + k] | diag2[i - k + N - 1]:
                continue
            col[i] = diag1[i + k] = diag2[i - k + N - 1] = True
            search(k + 1)
            col[i] = diag1[i + k] = diag2[i - k + N - 1] = False


N = int(sys.stdin.readline())
col = [False] * N
diag1 = [False] * (2 * N - 1)
diag2 = [False] * (2 * N - 1)
cnt = 0
search(0)
print(cnt)

File: 9663.cpp
#include <bits/stdc++.h>

using namespace std;

int n, cnt = 0;
vector <bool> col, diag1, diag2;

int search(int y)
{
	if (y == n) {
		cnt++;
	}
	else {
		for (int x = 0; x < n; x++) {
			if (col[x] || diag1[x + y] || diag2[x - y + n - 1])
				continue;
			col[x] = diag1[x + y] = diag2[x - y + n - 1] = true;
			search(y + 1);
			col[x] = diag1[x + y] = diag2[x - y + n - 1] = false;
		}
	}
	return 0;
}


int main()
{
	cin >> n;
	col.assign(n, false);
	diag1.assign(2 * n - 1, false);
	diag2.assign(2 * n - 1, false);
	search(0);
	cout << cnt;
	return 0;
}